lib/temporal/worker.rb

```diff
@@ -0,0 +1,71 @@
+require 'temporal/sdk/core/core_interface_pb'
+require 'temporal/bridge'
+require 'temporal/runtime'
+require 'temporal/worker/reactor'
+
+module Temporal
+  class Worker
+    class AlreadyStarted < StandardError; end
+
+    def initialize(connection, namespace, task_queue)
+      runtime = Temporal::Runtime.instance
+      runtime.ensure_callback_loop
+
+      @running = false
+      @core_worker = Temporal::Bridge::Worker.create(
+        runtime.core_runtime,
+        connection.core_connection,
+        namespace,
+        task_queue
+      )
+    end
+
+    def run
+      start
+      Reactor.attach
+    end
+
+    def start
+      raise AlreadyStarted if running?
+
+      @running = true
+
+      Reactor.execute do |reactor|
+        while running?
+          task = reactor.await { |r| poll_activity_task(r) }
+          reactor.async { |r| process_activity_task(task, r) }
+        end
+      end
+    end
+
+    def shutdown
+      @running = false
+    end
+
+    private
+
+    attr_reader :core_worker
+
+    def running?
+      @running
+    end
+
+    def poll_activity_task(resolver)
+      core_worker.poll_activity_task do |task|
+        decoded_task = Coresdk::ActivityTask::ActivityTask.decode(task)
+        resolver.call(decoded_task)
+      end
+    end
+
+    def process_activity_task(task, resolver)
+      puts "Processing #{task.start&.workflow_execution&.run_id} | #{task&.start&.activity_id}"
+
+      # TODO: Will be replaced with an actual processing login using a thread loop
+      Thread.new do
+        sleep 2
+        puts "Processed #{task.start&.workflow_execution&.run_id} | #{task.start&.activity_id}"
+        resolver.call
```

Comment:
What is resolver here? Why can't that start a thread if it needs to instead of assuming all activity tasks are threaded?

It is clearer to have the worker process activity events synchronously and have activity start be the thing that starts a thread if it needs to. You don't need to blindly use a thread for every activity task. For example, cancel doesn't need to run in a thread.

Reply:
`resolver` is just a block that we use to resolve a promise inside the reactor.

The whole `Thread.new` block here is just a placeholder for now and as per `TODO` comment will be replaced with putting an activity task processing on a thread pool.